cmd/kubeadm/app/util/cruntime.go

```diff
@@ -0,0 +1,141 @@
+*/
+
+package util
+
+import (
+	"fmt"
+	"strings"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	utilsexec "k8s.io/utils/exec"
+)
+
+type ContainerRuntimeType string
+
+// Supported runtime types
+const (
+	Docker ContainerRuntimeType = "docker"
+	CRI    ContainerRuntimeType = "cri"
+)
+
+// ContainerRuntime is a struct that interfaces with the container runtime
+type ContainerRuntime struct {
+	rtype ContainerRuntimeType
+	exec  utilsexec.Interface
+	tool  string
+	args  []string
+}
+
+// NewContainerRuntime sets up and returns a ContainerRuntime struct
+func NewContainerRuntime(execer utilsexec.Interface, criSocket string) (*ContainerRuntime, error) {
+	var err error
+	var toolName string
+	cruntime := ContainerRuntime{exec: execer}
+
+	if criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		cruntime.rtype = CRI
+		cruntime.args = []string{"-r", criSocket}
+		toolName = "crictl"
+	} else {
+		cruntime.rtype = Docker
+		toolName = "docker"
+	}
+
+	if cruntime.tool, err = execer.LookPath(toolName); err != nil {
+		return nil, fmt.Errorf("%s is required for %s container runtime: %v", toolName, cruntime.rtype, err)
+	}
+
+	return &cruntime, nil
+}
+
+// IsDocker returns true if the runtime is docker
+func (cruntime *ContainerRuntime) IsDocker() bool {
+	return cruntime.rtype == Docker
+}
+
+// IsRunning checks if runtime is running
+func (cruntime *ContainerRuntime) IsRunning() error {
+	if err := cruntime.Run("info"); err != nil {
+		return fmt.Errorf("container runtime is not running: %v", err)
+	}
+	return nil
+}
+
+func (cruntime *ContainerRuntime) getCommand(args ...string) utilsexec.Cmd {
+	cmdArgs := append(cruntime.args, args...)
+	return cruntime.exec.Command(cruntime.tool, cmdArgs...)
+}
+
+// Run creates command object and runs it to the completion
+func (cruntime *ContainerRuntime) Run(args ...string) error {
+	return cruntime.getCommand(args...).Run()
+}
+
+// CombinedOutput runs the command and returns its combined standard output
+// and standard error. This follows the pattern of package os/exec.
+func (cruntime *ContainerRuntime) CombinedOutput(args ...string) ([]byte, error) {
+	return cruntime.getCommand(args...).CombinedOutput()
+}
+
+// ListContainers lists running containers
+func (cruntime *ContainerRuntime) ListContainers() ([]string, error) {
+	var output []byte
+	var err error
+
+	if cruntime.rtype == CRI {
+		output, err = cruntime.CombinedOutput("pods", "-q")
+	} else {
+		output, err = cruntime.CombinedOutput("ps", "-a", "--filter", "name=k8s_", "-q")
+	}
+	return strings.Fields(string(output)), err
+}
+
+// RemoveContainers removes running containers
+func (cruntime *ContainerRuntime) RemoveContainers(containers []string) error {
+	var err, rcode error
+	rcode = nil
+	for _, container := range(containers) {
+		if cruntime.rtype == CRI {
+			err = cruntime.Run("stopp", container)
+			if err == nil {
+				err = cruntime.Run("rmp", container)
+			}
+		} else {
```

Comment:
better to do here `switch` with `cruntime.rtype == CRI', `cruntime.rtype == Docker` and `defaults:` with error.